src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs

```diff
@@ -157,6 +177,74 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
             : new BinaryArrayRecord(arrayInfo, memberTypeInfo);
     }
 
+    private static long GetJaggedArrayFlattenedLength(BinaryArrayRecord jaggedArrayRecord)
+    {
+        long result = 0;
+        Queue<BinaryArrayRecord>? jaggedArrayRecords = null;
+
+        do
+        {
+            if (jaggedArrayRecords is not null)
+            {
+                jaggedArrayRecord = jaggedArrayRecords.Dequeue();
+            }
+
+            Debug.Assert(jaggedArrayRecord.IsJagged);
+
+            foreach (object value in jaggedArrayRecord.Values)
+            {
+                object item = value is MemberReferenceRecord referenceRecord
+                    ? referenceRecord.GetReferencedRecord()
+                    : value;
+
+                if (item is not SerializationRecord record)
+                {
+                    result++;
+                    continue;
+                }
+
+                switch (record.RecordType)
+                {
+                    case SerializationRecordType.BinaryArray:
+                    case SerializationRecordType.ArraySinglePrimitive:
+                    case SerializationRecordType.ArraySingleObject:
+                    case SerializationRecordType.ArraySingleString:
+                        ArrayRecord nestedArrayRecord = (ArrayRecord)record;
+                        if (nestedArrayRecord.IsJagged)
+                        {
+                            (jaggedArrayRecords ??= new()).Enqueue((BinaryArrayRecord)nestedArrayRecord);
+                        }
+                        else
+                        {
+                            Debug.Assert(nestedArrayRecord is not BinaryArrayRecord, "Ensure lack of recursive call");
+                            checked
+                            {
+                                // In theory somebody could create a payload that would represent
+                                // a very nested array with total elements count > long.MaxValue.
+                                result += nestedArrayRecord.FlattenedLength;
+                            }
+                        }
+                        break;
```

Comment:
This isn't counting the arrays themselves.  Assuming that's intentional, you should have a comment at the top of the method (or before this switch) that explains what does, and what doesn't count.


```C#
object[][] objs = new object[][1_000_000];
objs.Fill(Array.Empty<object>());
```

Doesn't feel like it should have a smaller count than


```C#
object[][] objs = new object[][1_000_000];
objs.Fill(null);
```

And right now it feels like the first one says 0, and the second says a million.

Reply:
FWIW, I can understand not counting the arrays as answering how many `int` values might come out of an `int[][][]`; but then it gets weird when the `null`s get counted, because they're not `int` values.

So I guess there's a bit of "what is this number supposed to mean?".  The comment can explain that, and then the code can be judged against the comment.

Reply:
Good point, I've fixed that.